vim25/mo/registry.go

```diff
@@ -16,6 +13,28 @@ limitations under the License.
 
 package mo
 
-import "reflect"
+import (
+	"reflect"
+	"strings"
+)
 
 var t = map[string]reflect.Type{}
+
+func Add(name string, kind reflect.Type) {
+	t[name] = kind
+}
+
+type Func func(string) (reflect.Type, bool)
+
+func TypeFunc() Func {
```

Comment:
Is TypeFunc being used somewhere?  The vim25/types `TypeFunc` is used for xml decoding: https://github.com/vmware/govmomi/blob/50c576d6470e7ab9803ceb466ac34f936971dfc1/vim25/soap/client.go#L129

I think you just need the `Add()` func, so the vsan mo types can be inflated via PropertyCollector helpers, right?
But in that case, you wouldn't use the `vsan:` namespace prefix to register.  Can you clarify how your change is being used?

Reply:
I want to get all named fields from a given vmodl type name.
This is possible in rvc(rbvmomi). And we have similar requirements.

```
> type -h
usage: type name
Display information about a VMODL type
  name: VMODL type name
  -h, --help    Show this message

> type HostVsanSystem
Managed Object HostVsanSystem

Properties:
 config: VsanHostConfigInfo

Methods:
 AddDisks_Task(disk : HostScsiDisk[]) : VsanHostDiskMapResult
 EvacuateVsanNode_Task(maintenanceSpec : HostMaintenanceSpec, timeout : int) : void
 FetchVsanSharedSecret() : string
 InitializeDisks_Task(mapping : VsanHostDiskMapping[]) : VsanHostDiskMapResult
 QueryDisksForVsan(canonicalName : string[]) : VsanHostDiskResult
 QueryHostStatus() : VsanHostClusterStatus
 RecommissionVsanNode_Task() : void
 RemoveDiskMapping_Task(mapping : VsanHostDiskMapping[], maintenanceSpec : HostMaintenanceSpec) : VsanHostDiskMapResult
 RemoveDisk_Task(disk : HostScsiDisk[], maintenanceSpec : HostMaintenanceSpec, timeout : int) : VsanHostDiskResult
 UnmountDiskMapping_Task(mapping : VsanHostDiskMapping[]) : VsanHostDiskMapResult
 UpdateVsan_Task(config : VsanHostConfigInfo) : void
```

However, even with this commit, I still can't figure out a way to get **methods** of a given mo type.

Reply:
I think we'd need to generate more code to have a mapping from mo type to its method set, and the result type of each _Task method.  Not sure the best way to do that.  Don't want to mess with the existing generator, would prefer to replace it (#1186).  Also would not want to add more generated code to the existing packages with this data.  But would be ok with having in a separate package.  What do you think?

As for this change, the `mo` "registry" is used by methods such as `mo.ObjectContentToType` and `mo.LoadObjectContent`, which convert a PropertyCollector response into Go objects.  Those methods are not namespace aware, so I think you'd want to remove the `vsan:` prefix.

As for reading from the map, the vim25/types `TypeFunc` is used for xml decoding.  Maybe the mo package should have a more direct way to read from the map.  Note that you can currently read from it like so:
```go
	c := types.ObjectContent{Obj: types.ManagedObjectReference{Type: "Datastore"}}
	obj, _ := mo.ObjectContentToType(c)
	fmt.Printf("%T\n", obj)
```

Reply:
Thanks for your explanation! Your advice really helps.
Since we do not use property collector for vsan mo, this change is not necessary.
I also agree that we don't want addtional mappings to the existing package. I think I just need a generated map[string][]string.